python/car.py

```diff
@@ -38,4 +39,155 @@ def update_state(self, state):
         self.distance_taken = state[1]
         self.speed = state[2]
         self.acceleration_state = state[3]
+    
+    def get_follow_distance(self, safety_factor=1.0):
+        """
+        Distance traveled while getting to a full stop from current speed
+        :param safety_factor: returned value will be multiplied by this factor
+        """
+        # time to stop = speed / deceleration
+        # distance traveled = aree under the function of speed(time)
+        # which is a line from current speed at zero time, and zero speed at time to stop
+        follow_distance = (self.speed / 2.0) * (self.speed / self.specs.braking_power)
+        return safety_factor * follow_distance
+
+    def distance_while_accelerating(self, target_speed):
+        """
+        Distance traveled while reaching target_speed from current_speed
+        """
+        # we calculate the area under the function of speed(time) which is a trapeziod
+        if self.speed < target_speed:
+            return (target_speed + self.speed) / 2 * (target_speed - self.speed) / self.specs.acceleration
+        else:
+            return (target_speed + self.speed) / 2 * (self.speed - target_speed) / self.specs.braking_power
+
+    def match_speed_now(self,other_car):
+        #TODO not perfect, need a bit more calculations
+        return (self.distance_taken + self.match_speed_distance(other_car.speed) + 
+                other_car.get_follow_distance(safety_treshold=1.0) <
+                other_car.distance_taken)
+
+
+class CarManager(dict):
+    def __init__(self, *args):
+        dict.__init__(self, args)
+
+    def update_car(self, car_id, state):
+        self[car_id].update_state(state)
+
+
+# TODO: These should not be simple lists, but maybe something form bisect module
+class DetailedCarTracker(dict):
```

Comment:
Én úgy képzeltem el, hogy ez a kettő egymásból származik, tehát a CarManager az alap class, és a DetailedCarTracker a részletes. És meg lehet adni, hogy melyik classt használjuk hasonloan a model_class-hoz.
A dict-ből leszármazás szerintem nem is szükséges, csak legyenek implementálva __setitem__, __getitem__, get, pop metódusok mind a 2-re. A sima CarManagernek az adattagja lehetne a self.local_cars ami a dict maga